crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1601 @@
+                    // also means we can discard any potential errors from the `shutdown()`. Note that
+                    // `shutdown()` internally unwraps any critical errors.
+                    let _ = polars_io::pl_async::get_runtime()
+                        .block_on(self.try_shutdown())
+                        .unwrap();
+                };
+                self.is_finished.store(true, Ordering::Relaxed);
+            } else {
+                send[0] = PortState::Ready
+            }
+        }
+
+        fn spawn<'env, 's>(
+            &'env mut self,
+            scope: &'s TaskScope<'s, 'env>,
+            recv: &mut [Option<RecvPort<'_>>],
+            send: &mut [Option<SendPort<'_>>],
+            _state: &'s ExecutionState,
+            join_handles: &mut Vec<JoinHandle<PolarsResult<()>>>,
+        ) {
+            use std::sync::atomic::Ordering;
+
+            debug_assert!(recv.is_empty());
+            debug_assert_eq!(send.len(), 1);
+
+            let morsel_senders = send[0].take().unwrap().parallel();
+
+            #[cfg(debug_assertions)]
+            {
+                let guard = self.row_group_data_rx_handles.try_read().unwrap();
+                let (row_group_data_receivers, _) = guard.as_ref().unwrap();
+                assert_eq!(row_group_data_receivers.len(), morsel_senders.len());
+            }
+
+            self.row_group_data_task_init_permit.take();
+            let row_group_decoder = self.row_group_decoder.as_ref().unwrap().clone();
+
+            for (i, mut morsel_tx) in morsel_senders.into_iter().enumerate() {
+                let this = &*self;
+                let row_group_decoder = row_group_decoder.clone();
+                let row_group_data_rx_handles_guard =
+                    this.row_group_data_rx_handles.try_read().unwrap();
+
+                join_handles.push(scope.spawn_task(TaskPriority::Low, async move {
+                    let (row_group_data_receivers, _) =
+                        row_group_data_rx_handles_guard.as_ref().unwrap();
+                    let row_group_data_rx = &row_group_data_receivers[i];
+
+                    let wait_group = WaitGroup::default();
+                    let source_token = SourceToken::new();
+
+                    loop {
+                        let Ok(v) = row_group_data_rx.recv().await else {
+                            this.is_finished.store(true, Ordering::Relaxed);
+                            drop(row_group_data_rx_handles_guard);
+                            // Shutdown to propagate any errors.
+                            if let Some(v) = this.try_shutdown().await {
+                                v?;
+                            };
+                            break;
+                        };
+
+                        // Hold the permit until the downstream tasks are done with the morsel.
+                        let (handle, _row_group_data_fetch_permit) = WithSemaphorePermit::split(v);
```

Comment:
Why is `_row_group_data_fetch_permit` still a thing here? Shouldn't the backpressure from `row_group_data_rx.recv()` not already be enough? The sender should be blocked on sending until there is space available in the work distributor.

Reply:
It helps to enforce the limit precisely - without it we may fetch more than the configured prefetch size. I also just realized I forgot to update the code here - it's supposed to look something like this -

```rust
let row_group_data = {
    // Hold the permit until this task has finished fetching.
    let (handle, _row_group_data_fetch_permit) =
        WithSemaphorePermit::split(v);
    handle.await?
};
```

In a basic implementation, after a task handle is removed from the channel, there would be channel capacity and the producer would immediately spawn more tasks to fill the channel. However the problem with this is that for a channel of task handles, a task handle being popped by a consumer doesn't necessarily indicate that the task itself has completed, even though there is now channel capacity. By having the producer block on the semaphore instead of channel capacity, we ensure that the producer only spawns more tasks when a previous task has completed.

Although, I gave it some more thought and I think I can remove this permit and instead use a smaller channel size.

Reply:
I've removed it and instead we adjust the `num_receivers` and `buffer_size` of the distributor channel according to the prefetch size.

Reply:
I've updated the approach and made a new comment below.